### ISSUE TITLE
high memory usage

### ISSUE THREAD

**Issue by agauniyal** [richard1122/vscode-youcompleteme#11]
Last time I checked, it was making my system unusable, took 19.4% memory alone. I had to turn it off since then.

**Comment by agauniyal** [comment]
Also this message frequently appears, after which I've to restart vscode - 

> The ycm langauge server crashed 5 times in the last 3 minutes. The server will not be restarted.

```
events.js:154
      throw er; // Unhandled 'error' event
      ^

Error: connect ECONNREFUSED 127.0.0.1:40395
    at Object.exports._errnoException (util.js:890:11)
    at exports._exceptionWithHostPort (util.js:913:20)
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1061:14)
[Info  - 10:13:09 AM] Connection to server got closed. Server will restart.
events.js:154
      throw er; // Unhandled 'error' event
      ^

Error: connect ECONNREFUSED 127.0.0.1:45593
    at Object.exports._errnoException (util.js:890:11)
    at exports._exceptionWithHostPort (util.js:913:20)
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1061:14)
[Info  - 10:13:11 AM] Connection to server got closed. Server will restart.
events.js:154
      throw er; // Unhandled 'error' event
      ^

Error: connect ECONNREFUSED 127.0.0.1:33023
    at Object.exports._errnoException (util.js:890:11)
    at exports._exceptionWithHostPort (util.js:913:20)
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1061:14)
[Info  - 10:13:12 AM] Connection to server got closed. Server will restart.
events.js:154
      throw er; // Unhandled 'error' event
      ^

Error: connect ECONNREFUSED 127.0.0.1:40339
    at Object.exports._errnoException (util.js:890:11)
    at exports._exceptionWithHostPort (util.js:913:20)
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1061:14)
[Info  - 10:13:14 AM] Connection to server got closed. Server will restart.
events.js:154
      throw er; // Unhandled 'error' event
      ^

Error: connect ECONNREFUSED 127.0.0.1:38127
    at Object.exports._errnoException (util.js:890:11)
    at exports._exceptionWithHostPort (util.js:913:20)
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1061:14)
[Error - 10:13:15 AM] Connection to server got closed. Server will not be restarted.
```

**Comment by richard1122** [comment]
Could you please update to newest version.

Then config ycmd.debug true  in your configuration .

Is there anything important?

**Comment by richard1122** [comment]
May related to https://github.com/richard1122/vscode-youcompleteme/issues/10.
Sometimes ycmd may not killed success.

**Comment by agauniyal** [comment]
I've enabled the ycmd.debug conf to be true, and currently the plugin is eating about 24% memory. Where is the debug information present so I could post it here, I'm having slight difficulty finding them.

**Comment by agauniyal** [comment]
Here is another error from console of developer tools - 

```
workbench.main.js:121 Cannot read property 'activeParameter' of null: TypeError: Cannot read property 'activeParameter' of null
    at asSignatureHelp (/home/agauniyal/.vscode/extensions/RichardHe.you-complete-me-1.0.19/node_modules/vscode-languageclient/lib/protocolConverter.js:107:17)e.onUnexpectedError @ workbench.main.js:121
```

**Comment by richard1122** [comment]
I guess the extension may not correctly killed ycmd server. Could you please list your running programs ( ps -ef on linux, task manager on windows). Check are there many python process. Those should be killed after vscode close. You may manually close those ycmd servers.

**Comment by richard1122** [comment]
> Where is the debug information present so I could post it here, I'm having slight difficulty finding them.

You may found it in "Output".

Like:

![image](https://cloud.githubusercontent.com/assets/2534277/18668111/2439d914-7f66-11e6-94bd-892aaa35a376.png)

**Comment by richard1122** [comment]
I've released a newer version (1.0.20). May fixed this issue.

Could you please try it? 

Thanks.

**Comment by agauniyal** [comment]
I've updated it, the memory usage is above 25%. Here's the debug output - 

https://gist.github.com/agauniyal/fe539966a0a1c32a6f0a73e01a3dd297

**Comment by agauniyal** [comment]
I tried using vim and nvim and atom's autocomplete with my ycm installation. All of them are not crossing 3.5 - 4% of usage. This plugin starts out like that for some time, and then after a while it's suddenly at 19-20% of memory.

Would there any other information you would like to have. I'm willing to put some effort into solving this.

**Comment by agauniyal** [comment]
I tried using vim and nvim and atom's autocomplete with my ycm installation. All of them are not crossing 3.5 - 4% of usage. This plugin starts out like that for some time, and then after a while it's suddenly at 19-20% + of memory.

Would there any other information you would like to have. I'm willing to put some effort into solving this.

**Comment by agauniyal** [comment]
It increases dramatically whenever I hover over a variable to see it's type. Even if I'm hovering over a single var again and again, it's adding ~.5% each time 😮 .

Here's my ycm config - 

```
import os
import ycm_core

flags = [
'-Wall',
'-Wextra',
'-Werror',
'-fexceptions',
'-DNDEBUG',
'-std=c++14',
'-I/usr/include/c++',
]


compilation_database_folder = 'build'

if os.path.exists( compilation_database_folder ):
  database = ycm_core.CompilationDatabase( compilation_database_folder )
else:
  database = None

SOURCE_EXTENSIONS = [ '.cpp', '.cxx', '.cc', '.c' ]

def DirectoryOfThisScript():
  return os.path.dirname( os.path.abspath( __file__ ) )


def MakeRelativePathsInFlagsAbsolute( flags, working_directory ):
  if not working_directory:
    return list( flags )
  new_flags = []
  make_next_absolute = False
  path_flags = [ '-isystem', '-I', '-iquote', '--sysroot=' ]
  for flag in flags:
    new_flag = flag

    if make_next_absolute:
      make_next_absolute = False
      if not flag.startswith( '/' ):
        new_flag = os.path.join( working_directory, flag )

    for path_flag in path_flags:
      if flag == path_flag:
        make_next_absolute = True
        break

      if flag.startswith( path_flag ):
        path = flag[ len( path_flag ): ]
        new_flag = path_flag + os.path.join( working_directory, path )
        break

    if new_flag:
      new_flags.append( new_flag )
  return new_flags


def IsHeaderFile( filename ):
  extension = os.path.splitext( filename )[ 1 ]
  return extension in [ '.h', '.hxx', '.hpp', '.hh' ]


def GetCompilationInfoForFile( filename ):
  # The compilation_commands.json file generated by CMake does not have entries
  # for header files. So we do our best by asking the db for flags for a
  # corresponding source file, if any. If one exists, the flags for that file
  # should be good enough.
  if IsHeaderFile( filename ):
    basename = os.path.splitext( filename )[ 0 ]
    for extension in SOURCE_EXTENSIONS:
      replacement_file = basename + extension
      if os.path.exists( replacement_file ):
        compilation_info = database.GetCompilationInfoForFile(
          replacement_file )
        if compilation_info.compiler_flags_:
          return compilation_info
    return None
  return database.GetCompilationInfoForFile( filename )


# This is the entry point; this function is called by ycmd to produce flags for
# a file.
def FlagsForFile( filename, **kwargs ):
  if database:
    # Bear in mind that compilation_info.compiler_flags_ does NOT return a
    # python list, but a "list-like" StringVec object
    compilation_info = GetCompilationInfoForFile( filename )
    if not compilation_info:
      return None

    final_flags = MakeRelativePathsInFlagsAbsolute(
      compilation_info.compiler_flags_,
      compilation_info.compiler_working_dir_ )
  else:
    relative_to = DirectoryOfThisScript()
    final_flags = MakeRelativePathsInFlagsAbsolute( flags, relative_to )

  return {
    'flags': final_flags,
    'do_cache': True
  }
```

**Comment by richard1122** [comment]
@agauniyal  Will it increase memory usage on Atom with GetType (hover on vscode) ?

https://github.com/Qusic/atom-youcompleteme#keymaps

This is the key binding of Atom version, GetType is Alt+y t

**Comment by agauniyal** [comment]
@richard1122 yep that's it. around ~.3% on each call even with vim + ycm version. And I'm not even calling it on different variables, the same var. I should file an issue on ycmd repo.

Meanwhile, could you provide an config to disable this option? Please 😸

**Comment by richard1122** [comment]
@agauniyal This is very strange.. I've tested on my PC, after 1000 times GetType request, it only increased about 10MB memory usage. 😢 

Ok, I'll add an config.

**Comment by agauniyal** [comment]
I've the latest ycmd if that helps. And on my system.. 20% + means 800MB+ ...

Here's the repo I'm testing on (develop branch) - 
https://github.com/agauniyal/isaac-core/tree/develop/

**Comment by richard1122** [comment]
@agauniyal  I've released a newer version (1.0.21), now you can disable GetType temporarily.

**Comment by richard1122** [comment]
It looks like a ycmd related issue. I'll close it now. 
Please reopen it anytime if you need any help from vscode extension.
